### PR TITLE
Add take_while experimental dataset op

### PR DESCRIPTION
**Pull request by Squadrick**
Addresses #24105

### CODE REVIEW

**Comment by Squadrick** [comment]
Still need to add test cases, but any changes required in the current commits? 

@jsimsa @mrry Most other dataset op that take in a function as a parameter, check for short circuits using `ComputeShortCircuitIndices`, will I need to do that here as well?

Also, right now `take_while` is state-less, similar to `map`. Do I need to also build the `scan` equivalent for `take_while`, i.e. `take_while` with a `state`?

File: tensorflow/core/kernels/data/experimental/take_while_dataset_op.cc
Comment (jsimsa) [inline tensorflow/core/kernels/data/experimental/take_while_dataset_op.cc]: 2018 (or 2019)

File: tensorflow/core/kernels/data/experimental/take_while_dataset_op.cc
Comment (jsimsa) [inline tensorflow/core/kernels/data/experimental/take_while_dataset_op.cc]: The short-circuit logic is a performance optimization. Since this is an experimental op/kernel, you can leave it as a TODO (or implement it). I am fine either way.

File: tensorflow/core/kernels/data/experimental/take_while_dataset_op.cc
Comment (jsimsa) [inline tensorflow/core/kernels/data/experimental/take_while_dataset_op.cc]: This is not true. The cardinality of TakeWhileDataset will generally be unknown, so we should return `kUnknownCardinality` here.

Comment (jsimsa) [reply tensorflow/core/kernels/data/experimental/take_while_dataset_op.cc]: And because of this, we can probably get rid of the `preserve_cardinality` attribute, because cardinality might not be preserved even if the flag is set.

File: tensorflow/core/kernels/data/experimental/take_while_dataset_op.cc
Comment (jsimsa) [inline tensorflow/core/kernels/data/experimental/take_while_dataset_op.cc]: I think this might be true for the function passed to `map` but the contract for the function passed to `take_while` should be that early termination is indicated by returning `false`. So if the `take_while` predicate returns an error, I think it makes sense to simply propagate the error to the caller.

File: tensorflow/python/data/experimental/ops/take_while_ops.py
Comment (jsimsa) [inline tensorflow/python/data/experimental/ops/take_while_ops.py]: 2018 (or 2019)

File: tensorflow/python/data/experimental/ops/take_while_ops.py
Comment (jsimsa) [inline tensorflow/python/data/experimental/ops/take_while_ops.py]: you can just inline this in its single call site

**Review by jsimsa — CHANGES_REQUESTED** [review]
Overall, this looks promising. I left (mostly minor) comments throughout the PR.

As for your questions, 1) implementing the short-circuit path is optional and 2) `take_while` should be stateless. If a user needs a stateful version, they can build that out of existing transformations (`scan` + `take_while` + `map`).

File: tensorflow/core/kernels/data/experimental/take_while_dataset_op.cc
Comment (Squadrick) [reply tensorflow/core/kernels/data/experimental/take_while_dataset_op.cc]: I'll leave that as a TODO for now.

**Comment by Squadrick** [comment]
I've made the required changes. Adding tests in the next couple of hours.

**Comment by Squadrick** [comment]
@jsimsa I've added tests and made the changes mentioned. Take a look, and let me know if further changes are required.

File: tensorflow/core/kernels/data/experimental/take_while_dataset_op.cc
Comment (jsimsa) [inline tensorflow/core/kernels/data/experimental/take_while_dataset_op.cc]: nit: if `predicate(out_tensors)` returns `true`, the control-flow will reach here which means the comment "propage error to caller" is technically inaccurate because `s` could be `Status::OK()`.

Consider modifying lines 162 through 165 as follows:

```
*end_of_sequence = !bool_output[0].scalar<bool>()();
return Status::OK();
```

This will a) always explicitly set `end_of_sequence` and b) avoid the potentially misleading comment on line 167

File: tensorflow/python/data/experimental/kernel_tests/serialization/take_while_dataset_serialization_test.py
Comment (jsimsa) [inline tensorflow/python/data/experimental/kernel_tests/serialization/take_while_dataset_serialization_test.py]: Use a parameterized test for this. See https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/data/kernel_tests/batch_test.py for an example.

File: tensorflow/python/data/experimental/kernel_tests/take_while_test.py
Comment (jsimsa) [inline tensorflow/python/data/experimental/kernel_tests/take_while_test.py]: Use a parameterized test for this.

File: tensorflow/python/data/experimental/kernel_tests/take_while_test.py
Comment (jsimsa) [inline tensorflow/python/data/experimental/kernel_tests/take_while_test.py]: ditto

File: tensorflow/python/data/experimental/kernel_tests/take_while_test.py
Comment (jsimsa) [inline tensorflow/python/data/experimental/kernel_tests/take_while_test.py]: nit: rename "stringNotEquals" to "string_not_equals" or simply "not_equal"

**Review by jsimsa — CHANGES_REQUESTED** [review]
Thanks @Squadrick. Left a couple more small comments. I am going to trigger presubmit tests which might require additional fixes.

**Comment by Squadrick** [comment]
@jsimsa Made the changes and fixed whatever was failing on the tests. I wasn't exactly sure what the api_def should look like for this, so I copied and edited `api_def_FilterDataset`, so can you take a look and tell me whether it's right or wrong? 

Question: Also, what's the difference between `base_api` and `python_api`?

File: tensorflow/core/kernels/data/experimental/take_while_dataset_op.cc
Comment (jsimsa) [inline tensorflow/core/kernels/data/experimental/take_while_dataset_op.cc]: nit: include https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/util/ptr_util.h and then do `return MakeUnique<Iterator>({this, strings::StrCat(prefix, "::TakeWhile")}, loop_pred_)` here

File: tensorflow/core/kernels/data/experimental/take_while_dataset_op.cc
Comment (jsimsa) [inline tensorflow/core/kernels/data/experimental/take_while_dataset_op.cc]: replace `std::vector<Tensor>` with `const std::vector<Tensor>&`

File: tensorflow/python/data/experimental/kernel_tests/take_while_test.py
Comment (jsimsa) [inline tensorflow/python/data/experimental/kernel_tests/take_while_test.py]:
```
for value in boolean_array:
  if value:
    self.assertTrue(self.evaluate(next_element()))
  else:
    with self.assertRaises(errors.OutOfRangeError):
      self.evaluate(next_element())
    break;
```

and change `boolean_array` to:

```
values = [True]*size
values[index] = False 
```

where `size` and `0 <= index < size` are parameters of the test, using 3-5 combinations

**Review by jsimsa — CHANGES_REQUESTED** [review]
a couple more minor comments

File: tensorflow/python/data/experimental/kernel_tests/take_while_test.py
Comment (Squadrick) [reply tensorflow/python/data/experimental/kernel_tests/take_while_test.py]: I can instead do something like:

```python
boolean_array = [True]*size
boolean_array[index] = False

for _ in range(index):
  self.assertTrue(self.evaluate(next_element())

with self.assertRaises(errors.OutOfRangeError):
  self.evaluate(next_element())
```

**Comment by Squadrick** [comment]
@jsimsa Made the changes you asked and fixed whatever was causing the CI to fail. I think it should be good to go unless I've overlooked something.

**Review by jsimsa — DISMISSED** [review]

**Comment by Squadrick** [comment]
@jsimsa Can you help me out a bit?

Ubuntu Sanity fails on Step 5 of 12: `do_bazel_nobuild (bazel nobuild)`
```
ERROR: error loading package 'tensorflow/lite/experimental/swift': Extension file not found. 
Unable to load file '@bazel_skylib//lib:new_sets.bzl': file doesn't exist or isn't a file
```

Running`bazel build --nobuild  -- //tensorflow/... -//tensorflow/lite/java/demo/app/... -//tensorflow/lite/examples/android/... -//tensorflow/lite/schema/...` locally on `master` branch fails with the same warning for me. 

Is there something I'm missing, because I haven't edited those files at all.

**Comment by jsimsa** [comment]
@Squadrick can you try pull the `master` branch and then merging it into your PR? I am pretty sure the issue is unrelated to your change.

**Comment by Squadrick** [comment]
@jsimsa This is good to go, right? Nothing else to change?

**Comment by jsimsa** [comment]
Yes, I believe @ymodak is trying to get the presubmits to pass at which point it should move on to an internal review (should be a no-op) and then submitted.

**Comment by jsimsa** [comment]
@Squadrick one of the test failures `tensorflow/tools/api/tests:api_compatibility_test` is in fact related:

can you run the following:

```
    $ bazel build tensorflow/tools/api/tests:api_compatibility_test
    $ bazel-bin/tensorflow/tools/api/tests/api_compatibility_test \
          --update_goldens True
```

and update the PR

**Comment by Squadrick** [comment]
@jsimsa I pulled `master` and ran the commands. This is what my `git status` looks like now.
```
On branch take-while
Changes not staged for commit:
  (use "git add/rm <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   tensorflow/tools/api/golden/v2/tensorflow.data.experimental.pbtxt
	modified:   tensorflow/tools/api/golden/v2/tensorflow.estimator.-baseline-classifier.pbtxt
	deleted:    tensorflow/tools/api/golden/v2/tensorflow.estimator.-baseline-estimator.pbtxt
	modified:   tensorflow/tools/api/golden/v2/tensorflow.estimator.-baseline-regressor.pbtxt
	modified:   tensorflow/tools/api/golden/v2/tensorflow.estimator.-boosted-trees-classifier.pbtxt
	modified:   tensorflow/tools/api/golden/v2/tensorflow.estimator.-boosted-trees-regressor.pbtxt
	deleted:    tensorflow/tools/api/golden/v2/tensorflow.estimator.-checkpoint-saver-hook.pbtxt
	deleted:    tensorflow/tools/api/golden/v2/tensorflow.estimator.-checkpoint-saver-listener.pbtxt
	modified:   tensorflow/tools/api/golden/v2/tensorflow.estimator.-d-n-n-classifier.pbtxt
	deleted:    tensorflow/tools/api/golden/v2/tensorflow.estimator.-d-n-n-estimator.pbtxt
	modified:   tensorflow/tools/api/golden/v2/tensorflow.estimator.-d-n-n-linear-combined-classifier.pbtxt
	deleted:    tensorflow/tools/api/golden/v2/tensorflow.estimator.-d-n-n-linear-combined-estimator.pbtxt
	modified:   tensorflow/tools/api/golden/v2/tensorflow.estimator.-d-n-n-linear-combined-regressor.pbtxt
	modified:   tensorflow/tools/api/golden/v2/tensorflow.estimator.-d-n-n-regressor.pbtxt
	modified:   tensorflow/tools/api/golden/v2/tensorflow.estimator.-estimator.pbtxt
	deleted:    tensorflow/tools/api/golden/v2/tensorflow.estimator.-feed-fn-hook.pbtxt
	deleted:    tensorflow/tools/api/golden/v2/tensorflow.estimator.-final-ops-hook.pbtxt
	deleted:    tensorflow/tools/api/golden/v2/tensorflow.estimator.-global-step-waiter-hook.pbtxt
	modified:   tensorflow/tools/api/golden/v2/tensorflow.estimator.-linear-classifier.pbtxt
	deleted:    tensorflow/tools/api/golden/v2/tensorflow.estimator.-linear-estimator.pbtxt
	modified:   tensorflow/tools/api/golden/v2/tensorflow.estimator.-linear-regressor.pbtxt
	deleted:    tensorflow/tools/api/golden/v2/tensorflow.estimator.-logging-tensor-hook.pbtxt
	modified:   tensorflow/tools/api/golden/v2/tensorflow.estimator.-mode-keys.pbtxt
	deleted:    tensorflow/tools/api/golden/v2/tensorflow.estimator.-nan-loss-during-training-error.pbtxt
	deleted:    tensorflow/tools/api/golden/v2/tensorflow.estimator.-nan-tensor-hook.pbtxt
	deleted:    tensorflow/tools/api/golden/v2/tensorflow.estimator.-profiler-hook.pbtxt
	deleted:    tensorflow/tools/api/golden/v2/tensorflow.estimator.-second-or-step-timer.pbtxt
	deleted:    tensorflow/tools/api/golden/v2/tensorflow.estimator.-session-run-args.pbtxt
	deleted:    tensorflow/tools/api/golden/v2/tensorflow.estimator.-session-run-context.pbtxt
	deleted:    tensorflow/tools/api/golden/v2/tensorflow.estimator.-session-run-hook.pbtxt
	deleted:    tensorflow/tools/api/golden/v2/tensorflow.estimator.-session-run-values.pbtxt
	deleted:    tensorflow/tools/api/golden/v2/tensorflow.estimator.-step-counter-hook.pbtxt
	deleted:    tensorflow/tools/api/golden/v2/tensorflow.estimator.-stop-at-step-hook.pbtxt
	deleted:    tensorflow/tools/api/golden/v2/tensorflow.estimator.-summary-saver-hook.pbtxt
	deleted:    tensorflow/tools/api/golden/v2/tensorflow.estimator.experimental.-in-memory-evaluator-hook.pbtxt
	deleted:    tensorflow/tools/api/golden/v2/tensorflow.estimator.experimental.-linear-s-d-c-a.pbtxt
	deleted:    tensorflow/tools/api/golden/v2/tensorflow.estimator.experimental.pbtxt
	modified:   tensorflow/tools/api/golden/v2/tensorflow.estimator.pbtxt

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	diff
	tensorflow/tools/api/golden/v2/tensorflow.estimator.inputs.pbtxt

no changes added to commit (use "git add" and/or "git commit -a")
```

I also got this warning:
```
WARNING:tensorflow:Golden file update requested!
All test failures have been skipped, see the logs for detected diffs.
This test is now going to write new golden files.
Make sure to package the updates together with your change.

You will need an explicit API approval. This may take longer than a normal
review.
```

And this seems to be the `diff`:

```
+   member_method {
+     name: "take_while"
+     argspec: "args=[\'predicate\'], varargs=None, keywords=None, defaults=None"
+   }
```

I have no idea how to proceed. How do I go about getting API approval?

**Comment by jsimsa** [comment]
The only file `pbtxt` file you should add to your PR is `tensorflow/tools/api/golden/v2/tensorflow.data.experimental.pbtxt`. Ignore all the `tensorflow/tools/api/golden/v2/tensorflow.estimator.*` (I am not sure why they showed up).

I will take care of the API approval (it happens during the internal review of the PR).

**Review by jsimsa — APPROVED** [review]

File: tensorflow/tools/api/golden/v2/tensorflow.data.experimental.pbtxt
Comment (jsimsa) [inline tensorflow/tools/api/golden/v2/tensorflow.data.experimental.pbtxt]: Please re-run the update script and also include the `tensorflow/tools/api/golden/v`/tensorflow.data.experimental.pbtxt` file in your CL.

File: tensorflow/python/data/experimental/kernel_tests/serialization/BUILD
Comment (jsimsa) [inline tensorflow/python/data/experimental/kernel_tests/serialization/BUILD]: you are missing the `"@absl_py//absl/testing:parameterized"` dependency here

**Review by jsimsa — CHANGES_REQUESTED** [review]
The internal review surfaced a couple of issues.

File: tensorflow/tools/api/golden/v2/tensorflow.data.experimental.pbtxt
Comment (Squadrick) [reply tensorflow/tools/api/golden/v2/tensorflow.data.experimental.pbtxt]: Do you mean `tensorflow/tools/api/golden/v1/tensorflow.data.experimental.pbtxt`? I think there's a typo.

Comment (jsimsa) [reply tensorflow/tools/api/golden/v2/tensorflow.data.experimental.pbtxt]: yes

**Comment by Squadrick** [comment]
@jsimsa Running the update script didn't change any of the `v1` scripts.

In `tensorflow/api_template.__init__.py`, on line `56`, there's `_compat.enable_v2_behavior()`, this sets `_force_enable` in `tensorflow/python/tf2.py` to `True`, and every time `tf2.enabled()` is called it returns `True`. So setting `TF2_BEHAVIOR=0` has no effect on which TensorFlow version to run. `tensorflow/api_template.__init__.py` is one of the `srcs` for `api_compatibility_test`, so I had to manually remove `@test_util.run_v1_only('b/120545219')` in `api_compatibility_tests.py` on line `317` and `336`, to get v1 golden API files to update.

Is there something I'm overlooking?

**Comment by Squadrick** [comment]
@jsimsa Pushed the requested changes.

**Comment by jsimsa** [comment]
@Squadrick thanks, I will resume the internal review.

**Pull request merged**
